### PR TITLE
Bugfix/zcs 4021

### PR DESCRIPTION
**Pull request by thefoolwhocodes**
Implementation corresponding to ZCS 4021

### CODE REVIEW

**Comment by gordyt** [comment]
@thefoolwhocodes I didn't see anything in the pull request description, so quick question: Did you run this with [valgrind](http://valgrind.org/) to check for potential memory leaks?

File: thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c [hunk 1]
```diff
@@ -1043,6 +1054,174 @@ ngx_field_from_zmauthtoken(ngx_log_t *log, ngx_pool_t *pool,
...
+    rgc.err.data = errstr;
+
+
+    if (ngx_regex_compile(&rgc) != NGX_OK)
+    {
+        ngx_log_error(NGX_LOG_INFO, log, 0,"zmauth: ngx_claimdata_from_jwttoken regex compilation failed.Error=====%s",errstr);
+        return 0;
+    }
+
+    int captures[30];
```
Comment (plobbes) [inline thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c]: Looking at http://nginx.org/en/docs/dev/development_guide.html it appears the captures value should derive from ```rgc```? Is there a reason we're using 30 here? Based on my quick read, that value doesn't seem to correspond with the number of captures we're expecting (nor using). Am I missing something?

File: thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c [hunk 2]
```diff
@@ -1043,6 +1054,174 @@ ngx_field_from_zmauthtoken(ngx_log_t *log, ngx_pool_t *pool,
...
+            return 0;
+        }
+        ngx_log_debug1(NGX_LOG_DEBUG_ZIMBRA, log, 0, "zmauth: ngx_claimdata_from_jwttoken base64url decoded payload is %V",&urlDecodedPayload);
+
+        int ret = ngx_claimdata_from_payload(log,pool,&urlDecodedPayload,value);
+        return ret; 
+    }
+    else if (n == NGX_REGEX_NO_MATCHED)
+    {
+        /* no match was found */
```
Comment (plobbes) [inline thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c]: nit: remove superfluous comment (here and in block below, possibly even in the block above)

File: thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c [hunk 3]
```diff
@@ -1043,6 +1054,174 @@ ngx_field_from_zmauthtoken(ngx_log_t *log, ngx_pool_t *pool,
...
+        ngx_log_debug0(NGX_LOG_DEBUG_ZIMBRA, log, 0, "zmauth: ngx_claimdata_from_jwttoken no match found");
+        return 0;      
+    }
+    else
+    {
+        /* some error */
+        ngx_log_debug0(NGX_LOG_DEBUG_ZIMBRA, log, 0, "zmauth: ngx_claimdata_from_jwttoken no match found due to some error.");
```
Comment (plobbes) [inline thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c]: The return code (n) actually conveys info about the pcre_exec failure.
maybe try: ```..., "zmauth: ngx_claimdata_from_jwttoken match failed: %i", n...```

File: thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c [hunk 4]
```diff
@@ -1043,6 +1054,174 @@ ngx_field_from_zmauthtoken(ngx_log_t *log, ngx_pool_t *pool,
...
+    }
+
+    ngx_log_debug0 (NGX_LOG_DEBUG_HTTP, log, 0,
+            "zmauth: ngx_claimdata_from_jwttoken exit");
+ 
+    return 0;
+}
+
+/* extract sub data field from payload */
+static ngx_flag_t
```
Comment (plobbes) [inline thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c]: Similar comments as last function due to the duplication of code.

File: thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c [hunk 5]
```diff
@@ -1486,6 +1666,79 @@ zmauth_check_authtoken(ngx_http_request_t *r, ngx_str_t* field,
...
+    ngx_pool_t *pool;
+    ngx_log_t *log;
+    ngx_str_t token, value, *pvalue;
+    ngx_flag_t f;
+
+    pool = r->pool;
+    log = r->connection->log;
+
+    ngx_log_debug0 (NGX_LOG_DEBUG_HTTP, log, 0, "zmauth:zmauth_check_jwttoken starts");
+
```
Comment (plobbes) [inline thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c]: nit: remove extra blank line (and further below too)

File: thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c [hunk 6]
```diff
@@ -1486,6 +1666,79 @@ zmauth_check_authtoken(ngx_http_request_t *r, ngx_str_t* field,
...
+
+    //Bearer header extration starts
```
Comment (plobbes) [inline thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c]: nit: typo extraction

File: thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c [hunk 7]
```diff
@@ -1486,6 +1666,79 @@ zmauth_check_authtoken(ngx_http_request_t *r, ngx_str_t* field,
...
+    if (f) {
+        ngx_log_debug1 (NGX_LOG_DEBUG_HTTP, log, 0,
+                "zmauth:zmauth_check_jwttoken found ZM_JWT_TOKEN:%V",
+                &token);
+
+        f = ngx_claimdata_from_jwttoken(log, pool, &token, field, &value);
+
+
+        if (f) {
+            ngx_log_debug2 (NGX_LOG_DEBUG_HTTP, log, 0, "zmauth: got %V:%V from ZM_AUTH_TOKEN", field, &value);
```
Comment (plobbes) [inline thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c]: Is ```ZM_AUTH_TOKEN``` reference correct here?

File: thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c (hunk 1 above)
Comment (thefoolwhocodes) [reply thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c]: We can derive it from rgc.

File: thirdparty/nginx/zimbra-nginx/rpm/SPECS/nginx.spec [hunk 8]
```diff
@@ -17,6 +17,8 @@ URL:                http://nginx.org
 The Zimbra nginx build
 
 %changelog
+* Wed Apr 18 2018  Zimbra Packaging Services <packaging-devel@zimbra.com> - VERSION-1zimbra8.7b8ZAPPEND
+- ZCS-4021 for JWT support.
```
Comment (tmclane) [inline thirdparty/nginx/zimbra-nginx/rpm/SPECS/nginx.spec]: Is this the proper version to be setting this to?  Do we version the nginx updates independently from the version of ZCS they ship with?

File: thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c (hunk 3 above)
Comment (tmclane) [reply thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c]: 👍

File: thirdparty/nginx/zimbra-nginx/rpm/SPECS/nginx.spec (hunk 8 above)
Comment (thefoolwhocodes) [reply thirdparty/nginx/zimbra-nginx/rpm/SPECS/nginx.spec]: Yeah @tmclane  i checked with @Prashantsurana  and the version is fine.

**Comment by gordyt** [comment]
@thefoolwhocodes have you addressed the requests from @plobbes ?  I would like to get this merged and the ticket closed today if possible.

File: thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c (hunk 4 above)
Comment (thefoolwhocodes) [reply thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c]: Showing the error returned is done too as mentioned in the previous comment.

File: thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c [hunk 9]
```diff
@@ -1043,6 +1054,174 @@ ngx_field_from_zmauthtoken(ngx_log_t *log, ngx_pool_t *pool,
...
+        token.data += (sizeof("Bearer ") - 1);
+        token.len -= (sizeof("Bearer ") - 1);
+
+        ngx_log_debug1 (NGX_LOG_DEBUG_HTTP, log, 0, "zmauth: ngx_get_authheader_bearer Bearer token found:%V",&token);
+        *value = token;
+
+        return 1; 
+    }
+
+    ngx_log_debug0 (NGX_LOG_DEBUG_HTTP, log, 0,
```
[the remark this answers is not in the document]
Comment (thefoolwhocodes) [reply thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c]: Unwrapped the other longer lines

File: thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c (hunk 4 above)
Comment (thefoolwhocodes) [reply thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c]: @plobbes  can you please re review the changes done.

File: thirdparty/nginx/zimbra-nginx/rpm/SPECS/nginx.spec (hunk 8 above)
Comment (tmclane) [reply thirdparty/nginx/zimbra-nginx/rpm/SPECS/nginx.spec]: Thanks for checking!

File: thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c [hunk 10]
```diff
@@ -1043,6 +1048,192 @@ ngx_field_from_zmauthtoken(ngx_log_t *log, ngx_pool_t *pool,
...
+            "zmauth: ngx_get_authheader_bearer exit");
+
+    return 0;
+}
+
+/* extract claim data from jwt token */
+static ngx_flag_t
+ngx_claimdata_from_jwttoken(ngx_log_t *log, ngx_pool_t *pool,
+        ngx_str_t *authtoken, ngx_str_t *field, ngx_str_t *value) {
```
Comment (plobbes) [inline thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c]: nit: the opening curly brace in many (most?) cases seems to be on a line by itself, but here we're putting it at the end. There's far more inconsistency with this type of usage in this file than with the previous ``` if (...) { ```, but I figured I'd point it out.

File: thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c [hunk 11]
```diff
@@ -1043,6 +1048,192 @@ ngx_field_from_zmauthtoken(ngx_log_t *log, ngx_pool_t *pool,
...
+    ngx_memzero(&rgc, sizeof(ngx_regex_compile_t));
+
+    rgc.pattern  = pattern;
+    rgc.pool     = pool;
+    rgc.err.len  = NGX_MAX_CONF_ERRSTR;
+    rgc.err.data = errstr;
+    if (ngx_regex_compile(&rgc) != NGX_OK)
+    {
+        ngx_log_error(NGX_LOG_INFO, log, 0,
+                "zmauth: ngx_claimdata_from_payload regex compilation failed.Error:%s", errstr);
```
Comment (plobbes) [inline thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c]: nit: space after period (.)? - here and anywhere else we do something like this.

An alternative error might be: "zmauth: ngx_claimdata_from_payload regex compilation error=%s"

**Comment by thefoolwhocodes** [comment]
@plobbes  on requirements:
1. We depend upon the downstream server for validation.
2. @piyush4360  could you please add details on testing we carried out.

File: thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c [hunk 12]
```diff
@@ -1043,6 +1048,192 @@ ngx_field_from_zmauthtoken(ngx_log_t *log, ngx_pool_t *pool,
...
+
+    authValue = headers_in->authorization->value;
+
+    ngx_log_debug1 (NGX_LOG_DEBUG_HTTP, log, 0,
+            "zmauth: ngx_get_authheader_bearer token:%V", &authValue);
+
+    if (authValue.len >= sizeof("Bearer ") - 1
+            && ngx_memcmp(authValue.data,"Bearer ",sizeof("Bearer ")-1)
+            == 0)
+    {
```
[the remark this answers is not in the document]
Comment (piyush4360) [reply thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c]: Here are the links for test rail cases which contain the JWT testing done : 

http://testrail01.buf.synacor.com/testrail/index.php?/cases/view/537322
http://testrail01.buf.synacor.com/testrail/index.php?/cases/view/767249
http://testrail01.buf.synacor.com/testrail/index.php?/cases/view/767250
http://testrail01.buf.synacor.com/testrail/index.php?/cases/view/780141
http://testrail01.buf.synacor.com/testrail/index.php?/cases/view/820480
http://testrail01.buf.synacor.com/testrail/index.php?/cases/view/822721
http://testrail01.buf.synacor.com/testrail/index.php?/cases/view/903329
http://testrail01.buf.synacor.com/testrail/index.php?/cases/view/942598
http://testrail01.buf.synacor.com/testrail/index.php?/cases/view/1035935

**Pull request merged**